File: 1002.find-common-characters.js
/*
 * @lc app=leetcode id=1002 lang=javascript
 *
 * [1002] Find Common Characters
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {string[]}
 */
var commonChars = function(words) {
  words = words.map(word=>word.split('').sort().join('')).sort()
  console.log(words)

  let freq = {}
  words[0].split('').forEach(v=>{
    if (!freq[v]) freq[v] = 0
    freq[v]++
  })
  console.log(freq)

  for (let i=1; i<words.length; i++) {
    Object.entries(freq).forEach(([c, f])=>{
      for (let r=f; r>0; r--){
        if (!words[i].includes(c.repeat(r))) {
          freq[c]--
          if (freq[c] <= 0)
            delete freq[c]
        }
      }
    })
  }

  return Object.entries(freq).map(([c,f])=>c.repeat(f)).join('').split('')
};
// @lc code=end

File: 1002.find-common-characters.py
#
# @lc app=leetcode id=1002 lang=python3
#
# [1002] Find Common Characters
#

# @lc code=start
class Solution:
    def commonChars(self, words: List[str]) -> List[str]:
        freqHash = {}
        for c in words[0]:
            if c not in freqHash: freqHash[c] = 0
            freqHash[c] += 1
        for word in words[1:]:
            for c, num in list(freqHash.items()):
                freqHash[c] = min(word.count(c), num)
                if freqHash[c] == 0: freqHash.pop(c)

        result = []
        for k,n in freqHash.items():
            for i in range(0,n):
                result.append(k)
        return result
# @lc code=end

